File: python/n69_square.py
# https://leetcode-cn.com/problems/sqrtx/
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 1:
            return 1
        left = 0
        right = int(x/2)
        while left < right:
            mid = (left + right + 1) >> 1
            sqr = mid * mid
            if sqr <= x:
                left = mid
            else:
                right = mid - 1   
        return left
           
 
if __name__ == "__main__":
    solu = Solution()
    print(solu.mySqrt(1923514994))

File: cpp/n69_square.cpp
// https://leetcode-cn.com/problems/sqrtx/
#include <iostream>
using namespace std;
class Solution {
public:
    int mySqrt(int x) {
        if(x == 1) return 1;
        int left = 0;
        int right = x/2;
        while(left < right){
            long long mid = (left + right + 1) >> 1;
            if (mid * mid <= x) left = mid;
            else right = mid - 1;
        }
        return left;
    }
};

int main(){
    Solution solu;
    cout << solu.mySqrt(16) << endl;
}